File: Digital Business Enablement/buy.it-api/src/main/java/br/com/fiap/buy/it/service/DepartamentoService.java
package br.com.fiap.buy.it.service;

import br.com.fiap.buy.it.dto.DepartamentoDTO;
import br.com.fiap.buy.it.model.Departamento;
import br.com.fiap.buy.it.repository.DepartamentoRepository;

import br.com.fiap.buy.it.model.Tag;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class DepartamentoService {

    @Autowired
    private DepartamentoRepository departamentoRepository;

    @Autowired
    private TagService tagService;

    public Page<DepartamentoDTO> listAll(Pageable pageRequest) {
        return departamentoRepository.findAll(pageRequest).map(this::convertToDto);
    }

    public DepartamentoDTO findById(Long id) {
        Departamento departamento = findEntityById(id);
        return convertToDto(departamento);
    }

    public DepartamentoDTO create(DepartamentoDTO newData) {
        Departamento entity = convertToEntity(newData);
        Departamento savedEntity = departamentoRepository.save(entity);
        return convertToDto(savedEntity);
    }

    public DepartamentoDTO update(Long id, DepartamentoDTO updatedData) {
        findEntityById(id);
        updatedData.setId(id);
        Departamento updatedEntity = convertToEntity(updatedData);    
        Departamento savedEntity = departamentoRepository.save(updatedEntity);
        return convertToDto(savedEntity);
    }

    public void delete(Long id) {
        Departamento entity = findEntityById(id);
        departamentoRepository.delete(entity);
    }

    public Departamento findEntityById(Long id) {
        return departamentoRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Departamento não encontrado(a) por ID: " + id));
    }

    private DepartamentoDTO convertToDto(Departamento departamento) {
        DepartamentoDTO dto = new DepartamentoDTO();
        dto.setId(departamento.getId());
        dto.setNome(departamento.getNome());
        dto.setIcone(departamento.getIcone());
        if (departamento.getTags() != null) {
            Set<Long> idsTags = departamento.getTags().stream()
                    .map(Tag::getId)
                    .collect(Collectors.toSet());
            dto.setIdsTags(idsTags);
        }
        return dto;
    }

    private Departamento convertToEntity(DepartamentoDTO dto) {
        Departamento departamento;
        if (dto.getId() != null) {
            departamento = findEntityById(dto.getId());
            Set<Tag> newTags = new LinkedHashSet<>();
            if (dto.getIdsTags() != null) {
                dto.getIdsTags().forEach(id -> {
                    Tag tag = tagService.findEntityById(id);
                    newTags.add(tag);
                });
            }
            departamento.setTags(newTags);
        } else {
            departamento = new Departamento();
            departamento.setNome(dto.getNome());
            departamento.setIcone(dto.getIcone());
            if (dto.getIdsTags() != null) {
                dto.getIdsTags().forEach(id -> {
                    Tag tag = tagService.findEntityById(id);
                    departamento.addTag(tag);
                });
            }
        }
        return departamento;
    }
}

File: Enterprise Application Development/buyit/Services/DepartamentoService.cs
using Buyit.Context;
using Buyit.Dtos;
using Buyit.Models;
using Microsoft.EntityFrameworkCore;

namespace Buyit.Services;
public class DepartamentoService
{
    private readonly BuyitContext _context;

    public DepartamentoService(BuyitContext context)
    {
        _context = context;
    }

    public async Task<List<DepartamentoDto>> FindAll()
    {
        var list = await _context.Departamento
            .Include(x => x.Tags)
            .ToListAsync();

        return list.Select(entity => ConvertToDto(entity)).ToList();
    }

    public async Task<DepartamentoDto> FindById(long id)
    {
        var entity = await FindEntityById(id);
        return ConvertToDto(entity);
    }

    public async Task<DepartamentoDto> Create(DepartamentoDto newData)
    {
        var entity = await ConvertToEntity(newData);
        _context.Departamento.Add(entity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task<DepartamentoDto> Update(long id, DepartamentoDto updatedData)
    {
        var entity = await _context.Departamento
            .Include(x => x.Tags)
            .FirstOrDefaultAsync(x => x.Id == id);

        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}.");
        }

        entity.Tags.Clear();

        if (updatedData.IdsTags != null)
        {
            var newTags = await _context.Tag.Where(t => updatedData.IdsTags.Contains(t.Id)).ToListAsync();
            foreach (var tag in newTags)
            {
                entity.Tags.Add(tag);
            }
        }

        updatedData.Id = entity.Id;
        var updatedEntity = await ConvertToEntity(updatedData);
        _context.Entry(entity).CurrentValues.SetValues(updatedEntity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task Delete(long id)
    {
        var entity = await FindEntityById(id);
        _context.Departamento.Remove(entity);
        await _context.SaveChangesAsync();
    }

    public async Task<DepartamentoModel> FindEntityById(long id)
    {
        var entity = await _context.Departamento
            .Include(x => x.Tags)
            .FirstOrDefaultAsync(x => x.Id == id);
        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}");
        }
        return entity;
    }

    public async Task<List<DepartamentoDto>> FindByTagId(long id)
    {
        var list = await _context.Departamento
            .Where(d => d.Tags.Any(t => t.Id == id))
            .Include(x => x.Tags)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    private DepartamentoDto ConvertToDto(DepartamentoModel entity)
    {
        return new DepartamentoDto
        {
            Id = entity.Id,
            Nome = entity.Nome,
            Icone = entity.Icone,
            IdsTags = entity.Tags.Select(t => t.Id).ToList()
        };
    }

    private async Task<DepartamentoModel> ConvertToEntity(DepartamentoDto dto)
    {
        var tags = new List<TagModel>();
        if (dto.IdsTags != null && dto.IdsTags.Any())
        {
            tags = await _context.Tag.Where(tag => dto.IdsTags.Contains(tag.Id)).ToListAsync();
        }

        return new DepartamentoModel
        {
            Id = dto.Id ?? 0,
            Nome = dto.Nome,
            Icone = dto.Icone,
            Tags = tags
        };
    }
}